### PR TITLE
Layout tweaks

### PR DESCRIPTION
**Pull request by garrett**
I've made many layout changes, especially between 641px to 1200px wide.

Goals: 
1. Reduce the amount of padding
2. Increase the information density
3. Bring historical data up higher on the page whenever possible (main goal)

### CODE REVIEW

**Comment by garrett** [comment]
I'm not done with this yet, but I wanted to share what I have been working on... and also have a backup that's not just on my computer. :wink:

**Comment by martinpitt** [comment]
I landed #50 now.

**Comment by garrett** [comment]
Currently, the cards wrap oddly between:
- desktop mode: browser content 1095px (iframe content: 855px) down to 888px (iframe content: 648px)
- responsive mode: browser and iframe content 767px down to 648px

I guess I need to adjust the card width calc between 648px to 855px to make the cards 4-up? Or perhaps we can skim some from the top end if we can shrink the cards width further, especially after fixing the network list.

**Comment by garrett** [comment]
Additionally, the progress bars wrap oddly. I have a fix for this already in the media query, but I should expand when it's applied and make it affect mobile sizes too.

Screenshot:
Screenshot_2020-09-23 Performance Graphs - garrett Rain

How I have it handled in other breakpoints (when it needs more space):

Screenshot_2020-09-23 Performance Graphs - garrett Rain(1)

**Comment by garrett** [comment]
So, TODO:

- [ ] Fix wrapping in some breakpoints
- [ ] Fix progressbar alignment; make it apply to additional sizes

After I fix the above, and if you think this is good overall, it's probably ready for merging.

File: src/app.jsx [hunk 1]
```diff
@@ -789,7 +789,7 @@ class MetricsHistory extends React.Component {
         if (this.state.error)
             return <EmptyStatePanel
                         icon={ExclamationCircleIcon}
-                        title={_("Error has occured")}
+                        title={_("Error has occurred")}
```
Comment (martinpitt) [inline src/app.jsx:798]: This was already done in commit 492d303b8e50286aa0751256f8acdf7752c33909 . The other two typo fixes are straightforward, I pushed them to master. That fixes the rebase woes, on your next push you can drop this commit.

File: src/app.jsx [hunk 2]
```diff
@@ -450,11 +450,11 @@ class CurrentMetrics extends React.Component {
                     <CardBody>
                         <DescriptionList isHorizontal columnModifier={{ lg: '2Col' }}>
                             <DescriptionListGroup>
-                                <DescriptionListTerm>{ _("Reading") }</DescriptionListTerm>
+                                <DescriptionListTerm>{ _("Read") }</DescriptionListTerm>
```
Comment (martinpitt) [inline src/app.jsx:455]: This is fine. We shouldn't make a strong assumption about word length here as this gets translated, though.

I pushed this commit to master, as I want to help with some string/UI updates. So what remains here is the layout tweaks (which you are still working on).

Comment (garrett) [reply src/app.jsx:455]: > We shouldn't make a strong assumption about word length here as this gets translated

Agreed. That's why after I changed the text, I made sure the layout would stack the label and value in smaller sizes regardless.

**Comment by martinpitt** [comment]
@garrett: #69 landed now, so this isn't blocked on rebasing any more.

**Comment by martinpitt** [comment]
This is really cool! I like how the Network card makes itself really small vertically if you get into a 3-on-1 layout:

image

The only problem I noticed is with "Load", with a window width around 1450: there is no margin after the "Load:" label, and it breaks unhappily:

image

Now, we do format the "Load" text really sloppily. Would it help to render it differently, as three different spans or so? Or the whole thing as a DescriptionList? (although that by itself tends to claim a lot of space)

**Comment by martinpitt** [comment]
@garrett: Do you have some details about "Split out units from the network list into its own column"? Indeed this is a bit tricky, but not actually hard. I can send a draft PR. But wouldn't this cause ugly gaps between shorter numbers and their unit if there is also a longer number?

**Comment by martinpitt** [comment]
@garrett: Something like this?

image

(Note that this is still without the "near-zero numbers" fix in PR #71)

That's with this patch. I added empty column headers for the units.

```diff
--- src/app.jsx
+++ src/app.jsx
@@ -362,11 +362,12 @@ class CurrentMetrics extends React.Component {
         const num_cpu_str = cockpit.format(cockpit.ngettext("$0 CPU", "$0 CPUs", numCpu), numCpu);
         const have_storage = cockpit.manifests && cockpit.manifests.storage;
 
-        const netIO = this.netInterfacesNames.map((iface, i) => [
-            iface,
-            cockpit.format_bytes_per_sec(this.state.netInterfacesRx[i]),
-            cockpit.format_bytes_per_sec(this.state.netInterfacesTx[i]),
-        ]);
+        const netIO = this.netInterfacesNames.map((iface, i) => {
+            // we want number and unit as two separate columns
+            const fmt_rx = cockpit.format_bytes_per_sec(this.state.netInterfacesRx[i]).split(' ');
+            const fmt_tx = cockpit.format_bytes_per_sec(this.state.netInterfacesTx[i]).split(' ');
+            return [iface, fmt_rx[0], fmt_rx[1], fmt_tx[0], fmt_tx[1]];
+        });
 
         let swapProgress;
 
@@ -511,7 +512,7 @@ class CurrentMetrics extends React.Component {
                             gridBreakPoint={TableGridBreakpoint.gridLg}
                             borders={false}
                             aria-label={ _("Network usage") }
-                            cells={ [_("Interface"), _("In"), _("Out")] } rows={netIO}
+                            cells={ [_("Interface"), _("In"), "", _("Out"), ""] } rows={netIO}
                             rowWrapper={ props => <RowWrapper data-interface={ props.row[0] } {...props} /> }>
                             <TableHeader />
                             <TableBody />

```

**Comment by garrett** [comment]
@martinpitt: No, that doesn't actually reduce the amount of space, but increases it. But it does let us align the numbers better. I meant:


In | Out |  
-- | -- | --
123 | 555 | KiB/s

**Comment by martinpitt** [comment]
@garrett: Your table is missing the unit for "in". I'll look into having less precision there -- our `cockpit.format_bytes()` function does not allow that, but there must be some way. Also, my patch isn't good with the .split(), we should use the `separate` flag instead.

**Comment by garrett** [comment]
> Your table is missing the unit for "in".

Nope. They're both the same unit. That's why it's split out in its own row. (Why would you have in and out be different units?)

Also, perhaps we should put bounds on the results. Perhaps it should only be KiB/s at the lowest? 0 or 1 KiB/s is still extremely slow; B/s at this point isn't interesting, is it? The goal: Still attempting to make it easier to see what's going on.

**Comment by martinpitt** [comment]
> Why would you have in and out be different units

Because they are generally different.. Consider downloading a file, which will take lots of "In" (often several MB/s), but only a small "Out" (KB/s control stream). Most network operations are not very symmetric in terms of transmitting the same magnitude of data in both directions simultaneously. Videoconferencing may be an exception :-)

I'm very hesitant to declare that anything < 1 KB/s will just be shown as "0". There is a difference between a totally quiet interface, and one that still does receive traffic, like ssh or VPN keep-alives (a typical case for a steady stream of low-bandwidth traffic).

I do agree about reducing the precision, though -- there is no need for decimals.

**Comment by garrett** [comment]
> I'm very hesitant to declare that anything < 1 KB/s will just be shown as "0".

Then round up to 1? (That's what I was implying. I meant 0 as in actual 0. The difference between 0 and 1 is not much, hence 0.3 == 1 is not a big leap. Plus, it's based on an average, so it's never fully accurate anyway.) The point is to know that it's not much versus a lot... and having fake significant digits and granularity of B/s isn't useful; they get in the way more than anything.

Keeping it limited to KiB/s and MiB/s (and in some cases GiB/s???) is a lot easier to parse, as a human, than having to factor in B/s too — especially if numbers are not roundeed. Everyone's looking at the numbers first, not the units.

**Comment by marusak** [comment]
@garrett and @martinpitt what is state of this? I see you are having discussion about units and rounding off numbers. Is it something that should happen in this PR? Is it something I should take a look at?

**Comment by martinpitt** [comment]
@marusak: I think it should be a separate PR. We can't use `cockpit.format_*` for this, that's not flexible enough. But I'm sure stackoverflow/npmjs have some standard solution for that (or 500 :grin: ). If you have time to look into that, that'd be great!

**Comment by marusak** [comment]
>  I think it should be a separate PR

Ok. Lets open issue for that then. Anything else that needs to happen in this PR? (except rebase ofc).

**Comment by garrett** [comment]
> what is state of this?

Should be merged sooner than later.

> I see you are having discussion about units and rounding off numbers.

The network card has layout issues due to lack of rounding (but @martinpitt has a fix for that) and the units being duplicated for both in and out. We were talking about fixing this — somewhere — not necessarily in this PR, but it does affect the layout; hence talking about it here.

What I've done in this PR is shrink the font to be _way too small_ to compensate. It's a bad workaround, and the font size override should be removed once we get a proper fix (what we were talking about) implemented.

**Comment by marusak** [comment]
> Should be merged sooner than later.

Thanks for the info. I'll rebase, test it and create issue for those units. Lets move this forward as this block merging into cockpit.

**Review by marusak — APPROVED** [review]
Thanks!
I rebased but without actually doing any manual changes. Checked it out and it seems to work nicely.

**Pull request merged**